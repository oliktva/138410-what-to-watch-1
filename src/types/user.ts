import PropTypes from 'prop-types';

export const userPropTypes = PropTypes.shape({
  id: PropTypes.number,
  email: PropTypes.string,
  name: PropTypes.string,
  avatarUrl: PropTypes.string,
});

export interface UserProps {
  id?: number;
  email?: string;
  name?: string;
  avatarUrl?: string;
}
