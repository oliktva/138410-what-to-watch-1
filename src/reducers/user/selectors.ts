import namespaces from 'src/reducers/namespaces';

import {State} from 'src/types/reducer';
import {UserProps} from 'src/types/user';

export const getUser = (state: State): UserProps => {
  const {id, email, name, avatarUrl} = state[namespaces.USER];

  return {id, email, name, avatarUrl};
};

export const getError = (state: State): (string | undefined) => {
  return state[namespaces.USER].error;
};

export const getIsAuthorizationRequired = (state: State): boolean => {
  return state[namespaces.USER].isAuthorizationRequired;
};
