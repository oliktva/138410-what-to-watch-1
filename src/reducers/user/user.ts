import {Action as ReduxAction} from 'redux';
import {AxiosResponse, AxiosError, AxiosInstance} from 'axios';
import {toCamel} from 'convert-keys';

import {UserProps} from 'src/types/user';
import {ThunkDispatch, ThunkAction, State as AppState} from 'src/types/reducer';

export const TOGGLE_AUTHORIZATION_REQUIRED = `TOGGLE_AUTHORIZATION_REQUIRED`;
export const LOG_IN_USER = `LOG_IN_USER`;
export const LOG_IN_USER_ERROR = `LOG_IN_USER_ERROR`;

export interface State {
  isAuthorizationRequired: boolean;
  id?: number;
  email?: string;
  name?: string;
  avatarUrl?: string;
  error?: string;
}

interface LogInUserAction extends ReduxAction {
  type: typeof LOG_IN_USER;
  payload: UserProps;
}

interface LogInUserErrorAction extends ReduxAction {
  type: typeof LOG_IN_USER_ERROR;
  payload: string;
}

export type Action = LogInUserAction | LogInUserErrorAction;

export const initialState: State = {
  isAuthorizationRequired: false,
  id: undefined,
  email: undefined,
  name: undefined,
  avatarUrl: undefined,
  error: undefined
};

export const ActionCreator = {
  logInUser: (user: UserProps): LogInUserAction => {
    return {
      type: LOG_IN_USER,
      payload: user
    };
  },

  logInUserError: (error: string): LogInUserErrorAction => {
    return {
      type: LOG_IN_USER_ERROR,
      payload: error
    };
  }
};

export const Operation = {
  loadUser: (): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.get(`/login`)
        .then((response: AxiosResponse<Record<string, any>>): void => {
          const data = toCamel<UserProps>(response.data);

          dispatch(ActionCreator.logInUser(data));
        }).catch((): void => {
          // don't care
        });
    };
  },

  logInUser: (email: string, password: string): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.post(`/login`, {
        email,
        password
      }).then((response: AxiosResponse<Record<string, any>>): void => {
        const data = toCamel<UserProps>(response.data);

        dispatch(ActionCreator.logInUser(data));
      }).catch((error: AxiosError): void => {
        dispatch(ActionCreator.logInUserError(error.message));
      });
    };
  }
};

const reducer = (state: State = initialState, action: Action): State => {
  switch (action.type) {
    case LOG_IN_USER:
      return {
        ...state,
        ...action.payload,
        error: undefined
      };

    case LOG_IN_USER_ERROR:
      return {
        ...state,
        error: action.payload
      };

    default:
      return state;
  }
};

export default reducer;
