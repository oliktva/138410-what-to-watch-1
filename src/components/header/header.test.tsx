import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

import Header from './header';
import {user} from '../../fixtures/user';

Enzyme.configure({adapter: new Adapter()});

it(`renders correctly`, () => {
  const header = shallow(<Header user={user} className="test" />);

  expect(toJson(header)).toMatchSnapshot();
});
