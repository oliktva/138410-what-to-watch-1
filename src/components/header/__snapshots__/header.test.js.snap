// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<header
  className="page-header test"
>
  <Logo
    isLight={true}
  />
  <div
    className="user-block"
  >
    <Link
      to="/mylist"
    >
      <div
        className="user-block__avatar"
      >
        <img
          alt="User avatar"
          height="63"
          src="https://es31-server.appspot.comimg/1.png"
          width="63"
        />
      </div>
    </Link>
  </div>
</header>
`;
